lambda-buffers-compiler/src/LambdaBuffers/Compiler/ProtoCompat/Types.hs

```diff
@@ -44,142 +45,153 @@ module LambdaBuffers.Compiler.ProtoCompat.Types (
+ shouldn't be used anywhere outside of that
+-}
 data LBName = LBName {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary LBName
 
 data TyName = TyName {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary TyName
 
 data ConstrName = ConstrName {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary ConstrName
 
 data ModuleName = ModuleName {parts :: [ModuleNamePart], sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary ModuleName
 
 data ModuleNamePart = ModuleNamePart {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary ModuleNamePart
 
 data VarName = VarName {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary VarName
 
 data FieldName = FieldName {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary FieldName
 
 data ClassName = ClassName {name :: Text, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary ClassName
 
 newtype Kind = Kind {kind :: KindType}
   deriving stock (Show, Eq, Ord, Generic)
 
-data KindType
-  = KindRef KindRefType
-  | KindArrow Kind Kind
+instance Arbitrary Kind where
+  arbitrary = sized fn
+    where
+      fn n = Kind <$> resize n arbitrary
+
+data KindType = KindRef KindRefType | KindArrow Kind Kind
   deriving stock (Show, Eq, Ord, Generic)
 
-data KindRefType
-  = KUnspecified
-  | KType
+instance Arbitrary KindType where
+  arbitrary = sized fn
+    where
+      fn n
+        | n <= 0 = KindRef <$> arbitrary
+        | otherwise = KindArrow <$> resize (n `div` 2) arbitrary <*> resize (n `div` 2) arbitrary
+
+data KindRefType = KUnspecified | KType
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary KindRefType
 
 data TyVar = TyVar {varName :: VarName, sourceInfo :: SourceInfo}
   deriving stock (Show, Eq, Ord, Generic)
+  deriving (Arbitrary) via GenericArbitrary TyVar
 
-data Ty
-  = TyVarI TyVar
-  | TyAppI TyApp
-  | TyRefI TyRef
+data Ty = TyVarI TyVar | TyAppI TyApp | TyRefI TyRef
   deriving stock (Show, Eq, Ord, Generic)
 
-data TyApp = TyApp
-  { tyFunc :: Ty
-  , tyArgs :: NonEmpty Ty
-  , sourceInfo :: SourceInfo
-  }
+instance Arbitrary Ty where
+  arbitrary = sized fn
+    where
+      fn :: (Num a, Ord a) => a -> Gen Ty
+      fn n
+        | n <= 0 = TyRefI <$> arbitrary
+        | otherwise =
+            oneof
+              [ TyVarI <$> arbitrary
+              , TyAppI <$> arbitrary
+              , TyRefI <$> arbitrary
+              ]
```

Comment:
How do you feel about keeping instances close to their type definitions. Imo it bloats a module that should be a pure type lib that we maintain as a highly readable place. I'd rather have a separate module where the Arbitrary instances are orphaned.

Reply:
I guess it's a matter of taste. I like having instances next to their definition - I find it's the nicest way especially when navigating library code. There's a second option where they go after the type definitions.

I'm not a fan of orphaned definitions - to me it _feels_ wrong. I'm also unsure how that affects the haddock output. 

I'd keep them in the file - if you want we can move them at the bottom. I would not make them orphan.